### PR TITLE
test: support upvalues in exec_lua

### PR DESCRIPTION
**Pull request by lewis6991**

### CODE REVIEW

File: test/functional/testnvim.lua [hunk 1]
```diff
@@ -835,19 +835,126 @@ function M.exec_capture(code)
...
+      break
+    end
+    upvalues[n] = v
+    i = i + 1
+  end
+
+  local invalid_types = {
+    ['thread'] = true,
+    ['function'] = true,
+    ['userdaya'] = true,
```
Comment (tomtomjhj) [inline test/functional/testnvim.lua:908]:
```suggestion
    ['userdata'] = true,
```

File: test/functional/lua/buffer_updates_spec.lua [hunk 2]
```diff
@@ -282,19 +282,19 @@ describe('lua buffer event callbacks: on_lines', function()
...
+    local code = function()
       local buf = vim.api.nvim_create_buf(false, false)
 
-      vim.cmd"split"
+      vim.cmd 'split'
       vim.api.nvim_win_set_buf(0, buf)
 
       vim.api.nvim_buf_attach(buf, false, {
-        on_detach = function(_, buf)
+        on_detach = function(_, buf0)
```
Comment (justinmk) [inline test/functional/lua/buffer_updates_spec.lua:292]: Can you say a bit about what's going on here? `buf0` was renamed in some places, but not others.

File: test/functional/lua/filetype_spec.lua [hunk 3]
```diff
@@ -152,7 +152,12 @@ describe('vim.filetype', function()
       xml = { formatexpr = 'xmlformat#Format()' },
     } do
       for option, value in pairs(opts) do
-        eq(value, exec_lua([[ return vim.filetype.get_option(...) ]], ft, option))
+        eq(
+          value,
+          exec_lua(function()
+            return vim.filetype.get_option(ft, option)
```
Comment (justinmk) [inline test/functional/lua/filetype_spec.lua:158]: Another advantage of non-string `exec_lua` is that stylua can format it.

File: test/functional/testnvim.lua [hunk 4]
```diff
@@ -835,19 +835,156 @@ function M.exec_capture(code)
...
+  assert(session)
+
+  if not session.exec_lua_setup then
+    M.api.nvim_exec_lua(
       [[
-      local code = ...
-      return loadstring(code)(select(2, ...))
+      _G.__get_upvalues = loadstring((select(1,...)))
+      _G.__set_upvalues = loadstring((select(2,...)))
+      _G.__exec_lua_handler = loadstring((select(3,...)))
```
Comment (justinmk) [inline test/functional/testnvim.lua:963]: nit: store everything in a single `_G.__test` object?

```suggestion
      _G.__test = {
        get_upvalues = loadstring((select(1,...))),
        set_upvalues = loadstring((select(2,...))),
        exec_lua_handler = loadstring((select(3,...))),
      }
```

Comment (lewis6991) [reply test/functional/testnvim.lua]: done

**Comment by justinmk** [comment]
Even though I was hesitant about this, it's clear now that it gains 3+ improvements to clarity and robustness. Thanks for seeing it through @lewis6991

File: test/functional/testnvim.lua [hunk 5]
```diff
@@ -837,19 +837,171 @@ function M.exec_capture(code)
...
+    end
+    i = i + 1
+  end
+end
+
+--- @type fun(f: function): table<string,any>
+_G.__get_upvalues = nil
+
+--- @type fun(f: function, upvalues: table<string,any>)
+_G.__set_upvalues = nil
```
Comment (justinmk) [inline test/functional/testnvim.lua:876]: Not used anymore?

**Pull request merged**